src/response/errors.rs

```diff
@@ -49,30 +49,33 @@ struct ErrorMsg {
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match *self {
-            Error::Unsupported => write!(f, "Unsupported Operation"),
-            Error::Unauthorized => write!(f, "Authorization required"),
-            Error::BlobUnknown => write!(f, "Blob Unknown"),
+            Error::Unsupported => format_error_json(f, "UNSUPPORTED".to_string(), "Unsupported".to_string(), json!({})),
+            Error::Unauthorized => format_error_json(f, "UNAUTHORIZED".to_string(), "Authorization required".to_string(), json!({})),
+            Error::BlobUnknown => format_error_json(f, "BLOB_UNKNOWN".to_string(), "Blob Unknown".to_string(), json!({})),
             Error::BlobUploadUnknown => write!(f, "Blob Upload Unknown"),
-            Error::InternalError => write!(f, "Internal Error"),
-            Error::DigestInvalid => write!(f, "Provided digest did not match uploaded content"),
-            Error::ManifestInvalid => write!(f, "Manifest Invalid"),
-
-            Error::ManifestUnknown(ref tag) => {
-                let emsg = ErrorMsg {
-                    code: "MANIFEST_UNKNOWN".to_string(),
-                    message: "manifest unknown".to_string(),
-                    detail: json!({ "Tag": tag }),
-                };
-                write!(
-                    f,
-                    "{{\"errors\":[{}]}}",
-                    serde_json::to_string(&emsg).unwrap_or("MANIFEST_UNKNOWN".to_string())
-                )
-            }
+            //TODO: INTERNAL_ERROR code is not in the distribution spec
+            Error::InternalError => format_error_json(f, "INTERNAL_ERROR".to_string(), "Internal Server Error".to_string(), json!({})),
+            Error::DigestInvalid => format_error_json(f, "DIGEST_INVALID".to_string(), "Provided digest did not match uploaded content".to_string(), json!({})),
+            Error::ManifestInvalid => format_error_json(f, "MANIFEST_INVALID".to_string(), "manifest invalid".to_string(), json!({})),
+            Error::ManifestUnknown(ref tag) => format_error_json(f, "MANIFEST_UNKNOWN".to_string(), "manifest unknown".to_string(), json!({ "Tag": tag }))
         }
     }
 }
 
+fn format_error_json(f: &mut fmt::Formatter, code: String, message: String, detail: serde_json::Value) -> fmt::Result {
+    let emsg = ErrorMsg {
+        code: code,
+        message: message,
+        detail: detail,
```

Comment:
I wonder if we should change the detail to be `Option<String>`.

Reply:
I didn't do it just because I didn't know how the None would be translated to json. I can check it out.

Reply:
I wouldn't do that because then we are pushing responsiblity to generate proper json on the caller of the function.

Reply:
ideally, we could pass arbitrary struct here with details and that would be translated to json string, but I don't know how can I do that.

Reply:
I'm not sure what you mean. I'm pretty sure it will all just work if you just change:
```
struct ErrorMsg {
    code: String,
    message: String,
    detail: Value,
}
```
to:

```
struct ErrorMsg {
    code: String,
    message: String,
    detail: Option<Value>,
}
```
Then if there's no details, you can just pass None. I think serde will still happily deal with converting to Json.

Reply:
please take a look now. It that, what you meant?